Add Docker Install Docs

JONBRWN opened the pull request:
**Add Docker Install Docs**

- Adds Docker Setup Instructions and "Run A Application in Docker". 
- Adds these new docs under "First Install" within "Getting Started". 
- Moves existing documentation under "Installing From Source" within "Getting Started". 
- Updates Python example README's with `Note` comment for running in Docker.

**Files to Review:**
- [Docker Setup](https://github.com/WallarooLabs/wallaroo/blob/add-docker-install-docs/book/getting-started/docker-setup.md)
- [Run A Application in Docker](https://github.com/WallarooLabs/wallaroo/blob/add-docker-install-docs/book/getting-started/run-a-wallaroo-application-docker.md)
- All README's in the [examples/python](https://github.com/WallarooLabs/wallaroo/tree/add-docker-install-docs/examples/python) directory.
- [Tips for running Wallarooin Docker](https://github.com/WallarooLabs/wallaroo/blob/add-docker-install-docs/book/appendix/wallaroo-in-docker-tips.md)
- [Choosing an Installation Option for Wallaroo](https://github.com/WallarooLabs/wallaroo/blob/add-docker-install-docs/book/getting-started/choosing-an-installation-option.md)


Closes: #1681 #1682 #1683

dipinhora commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,154 @@
+
+We'll be naming our Docker container `wally` so we can access it with convenience and we'll be exposing port 4000 so we can view the Metrics UI. This shell must remain running in order for the container to be kept alive.
+
+Note: For each additional shell needed, run the following command:
```
It seems likely that people will not connect your reference of `additional shell` with the `Terminal N` later on and will forget to run the `docker exec` command. It might be better to to be explicit for each new `Terminal N` required and have them run the `docker exec` as the first command in each.

JONBRWN replied:
@dipinhora this has been addressed

dipinhora commented on examples/python/alphabet_partitioned/README.md:
```diff
@@ -34,7 +34,9 @@ The `Decoder`'s `decode(...)` method creates a `Votes` object with the letter be
 ## Running Alphabet Partitioned
 
-In order to run the application you will need Machida, Giles Sender, Giles Receiver, and the Cluster Shutdown tool. To build them, please see the [Linux](/book/getting-started/linux-setup.md) or [Mac OS](/book/getting-started/macos-setup.md) setup instructions.
+In order to run the application you will need Machida, Giles Sender, Giles Receiver, and the Cluster Shutdown tool. To build them, please see the [Linux](/book/getting-started/linux-setup.md) or [Mac OS](/book/getting-started/macos-setup.md) setup instructions. Alternatively, they could be run in Docker, please see the [Docker](/book/getting-started/docker-setup.md) setup instructions.
```
Why are we telling people to go look at the docker installation instructions here and in other similar places? i would have expected to send them to the page explaining how to run a wallaroo app in docker.....

JONBRWN replied:
@dipinhora this has been addressed, let me know if it makes more sense now

JONBRWN commented:
@nisanharamati when you review, can you check whether the virtualenv information is enough for the user to understand in conjunction with your new virtualenv documentation or whether additional information needs to be added that is docker specific.

nisanharamati commented on SUMMARY.md:
```diff
@@ -13,11 +13,14 @@
-* [Run a Wallaroo Application](book/getting-started/run-a-wallaroo-application.md)
+* First Install
+  * [Setting Up Your Environment](book/getting-started/docker-setup.md)
+  * [Run a Wallaroo Application in Docker](book/getting-started/run-a-walllaroo-application-docker.md)
```
Bad link, too many 'l's in "wallaroo"

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+# Setting Up Your Environment for Wallaroo in Docker
+
+To get you up and running quickly with Wallaroo, we've provided a Docker image which includes Wallaroo and related tools needed to run and modify a few example applications. We should warn that this Docker image was created with the intent of getting users started quickly with Wallaroo and is not intended to be a fully customizable development environment or suitable for production.
```
nit: "we have" instead of "we've". Contractions read weird in official docs.

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+### MacOS
+
+There are [instructions](https://docs.docker.com/docker-for-mac/) for getting Docker up and running on MacOS on the [Docker website](https://docs.docker.com/docker-for-mac/).  We recommend the 'Standard' version of the 'Docker for Mac' package.
```
impeartive voice is preferable here: "Follow the instructions on [the Docker website](...)..."

JONBRWN replied:
some of these tone comments I won't address as they are the same across a few existing files. Don't want to change them all as part of this PR but also don't want to change it in a single place so we're carrying different tones throughput.

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+There are [instructions](https://docs.docker.com/docker-for-mac/) for getting Docker up and running on MacOS on the [Docker website](https://docs.docker.com/docker-for-mac/).  We recommend the 'Standard' version of the 'Docker for Mac' package.
+
+Installing Docker will result in it running on your machine. After you reboot your machine, that will no longer be the case. In the future, you'll need to have Docker running in order to use a variety of commands in this book. We suggest that you [set up Docker to boot automatically](https://docs.docker.com/docker-for-mac/#general).
```
"At the end of the installation, the installer starts the docker daemon. After rebooting, however, the daemon is not started automatically.  If you want to have docker start automatically, we suggest [these instructions](...). If you prefer to leave it off by default, you will need to manually start the docker daemon before using it, with the following command ..."

JONBRWN replied:
same as above

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+### Linux Ubuntu
+
+There are [instructions](https://docs.docker.com/engine/installation/linux/ubuntu/) for getting Docker up and running on Ubuntu on the [Docker website](https://docs.docker.com/engine/installation/linux/ubuntu/).
```
Same note about using the imperative voice as above in the MacOS section

JONBRWN replied:
same as above

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+There are [instructions](https://docs.docker.com/engine/installation/linux/ubuntu/) for getting Docker up and running on Ubuntu on the [Docker website](https://docs.docker.com/engine/installation/linux/ubuntu/).
+
+Installing Docker will result in it running on your machine. After you reboot your machine, that will no longer be the case. In the future, you'll need to have Docker running in order to use a variety of commands in this book. We suggest that you [set up Docker to boot automatically](https://docs.docker.com/engine/installation/linux/linux-postinstall/#configure-docker-to-start-on-boot).
```
Same note as the one in the MacOS section.

JONBRWN replied:
same as above

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+Installing Docker will result in it running on your machine. After you reboot your machine, that will no longer be the case. In the future, you'll need to have Docker running in order to use a variety of commands in this book. We suggest that you [set up Docker to boot automatically](https://docs.docker.com/engine/installation/linux/linux-postinstall/#configure-docker-to-start-on-boot).
+
+All of the Docker commands throughout the rest of this manual assume that you have permission to run Docker commands as a non-root user. Follow the [Manage Docker as a non-root user](https://docs.docker.com/engine/installation/linux/linux-postinstall/#manage-docker-as-a-non-root-user) instructions to set that up. If you don't want to allow a non-root user to run Docker commands, you'll need to ru ...
```
"The commands throughout..."
No need for "All of the Docker" here.

"As a non-root user" --> "without root permissions"
"You can [follow this guide] to set up docker for use without root, or otherwise use `sudo` when issuing Docker commands.

JONBRWN replied:
same as above

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+All of the Docker commands throughout the rest of this manual assume that you have permission to run Docker commands as a non-root user. Follow the [Manage Docker as a non-root user](https://docs.docker.com/engine/installation/linux/linux-postinstall/#manage-docker-as-a-non-root-user) instructions to set that up. If you don't want to allow a non-root user to run Docker commands, you'll need to ru ...
+
+## Pull the official Wallaroo image from [Bintray](https://bintray.com/wallaroo-labs/wallaroolabs/first-install%3Awallaroo):
```
language nit: "Get the Official..." rather than "pull".
`pull` is the docker command to get something from a remote. But the English description is of "getting something from somewhere".

nisanharamati commented on book/getting-started/docker-setup.md:
````diff
@@ -0,0 +1,55 @@
+docker pull wallaroo-labs-docker-wallaroolabs.bintray.io/dev/wallaroo:7b1f3550
+```
+
+## What's Included
````
in what? 

Also, I think this could do without addition section heading. Just a sentence following the pull command:

"Included in the image are:"

and then a bullet list of the items in the image:

```
* Machida
    * machida description
* Giles Sender
    * giles description
...
```

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+## Conclusion
+
+Awesome! All set. Time to try running your first Wallaroo application in Docker.
```
I'd remove the "Awesome!", to keep the tone consistent with the rest of the docs, and add a link to the "running your first wallaroo app in docker" page.

aturley replied:
I'll point out that "Awesome" appears in a number of places in the docs. While I'm not a big fan of having it anywhere in the documentation, it isn't inconsistent with the tone in other places.

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+# Run a Wallaroo Application in Docker
+
+In this section, we're going to run an example Wallaroo application in Docker. By the time you are finished, you'll have validated that your Docker environment is set up and working correctly. If you haven't already completed the Docker setup [instructions](book/getting-started/docker-setup.md), please do so before continuing.
```
"we're going to" -> "we will"

nisanharamati replied:
"you'll" -> "you will"

nisanharamati replied:
"By the time you are finished," -> "By the time we have finished,"

Note that the active voice here is "we", as in, us the authors PLUS you the reader. So the learning is done by "you", but the actions are all done by "we"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+In this section, we're going to run an example Wallaroo application in Docker. By the time you are finished, you'll have validated that your Docker environment is set up and working correctly. If you haven't already completed the Docker setup [instructions](book/getting-started/docker-setup.md), please do so before continuing.
+
+There's a couple Wallaroo support applications that you'll be interacting with for the first time:
```
"You will be interacting with a few Wallaroo auxiliary applications for the first time:"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+There's a couple Wallaroo support applications that you'll be interacting with for the first time:
+
+- Our Metrics UI that allows you to monitor the performance and health of your applications.
```
"Metrics UI, which allows..." (drop the "our")

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+- Our Metrics UI that allows you to monitor the performance and health of your applications.
+- Giles receiver is designed to capture TCP output from Wallaroo applications.
+- Giles sender is used to send test data into Wallaroo applications over TCP.
+- Machida, our program for running Wallaroo Python applications.
```
"Machida runs Wallaroo Python applications."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+- Giles sender is used to send test data into Wallaroo applications over TCP.
+- Machida, our program for running Wallaroo Python applications.
+
+You're going to set up our "Celsius to Fahrenheit" example application. Giles sender will be used to pump data into the application. Giles receiver will receive the output and our Metrics UI will be running so you can observe the overall performance.
```
`We will set up the "Celsius to Fahrenheit" example application. We will use Giles Receiver for the sink, Giles Sender to send input data to the source, and the Metrics UI to observe the overall performance.`

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+You're going to set up our "Celsius to Fahrenheit" example application. Giles sender will be used to pump data into the application. Giles receiver will receive the output and our Metrics UI will be running so you can observe the overall performance.
+
+The Metrics UI process will be run in the background. The other three processes (receiver, sender, and Wallaroo) will run in the foreground. We recommend that you run each process in a separate terminal.
```
`Metrics UI runs in the background, while the receiver, sender, and Wallaroo will run in the foreground. We recommend running each process in a separate terminal.`

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+The Metrics UI process will be run in the background. The other three processes (receiver, sender, and Wallaroo) will run in the foreground. We recommend that you run each process in a separate terminal.
+
+NOTE: You might need to run Docker commands with sudo depending on how you set up Docker.
```
NOTE: if you haven't [set up docker to run without root], you will need to use `sudo` with your Docker commands.

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+NOTE: You might need to run Docker commands with sudo depending on how you set up Docker.
+
+Let's get started!
```
I think this can be omitted

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+## Terminal 1, Start the Wallaroo Docker container
+
+```bash
+docker run --rm -it --privileged -p 4000:4000 -v /tmp/wallaroo-docker/wallaroo-src:/src/wallaroo -v /tmp/wallaroo-docker/python-virtualenv:/src/python-virtualenv --name wally wallaroo-labs-docker-wallaroolabs.bintray.io/dev/wallaroo:7b1f3550
````
Can you break this line up with `\` so it doesn't have the side scroll and all of the text is fully visible?

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+### Breaking down the Docker command
+
+- `docker run`: The Docker command that will start a new container.
```
"The Docker command to start a new container."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `docker run`: The Docker command that will start a new container.
+
+- `--rm`: This flag will automatically clean up the container and remove the file system on exit.
```
You can omit the 'this flag will' part from all of these.

nisanharamati replied:
so the description changes from "This flag will automatically..." to "Automatically..."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `--rm`: This flag will automatically clean up the container and remove the file system on exit.
+
+- `-it`: This flag will allow us to work with interactive processes by allocating a tty for the container.
```
omit "this flag will"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `-it`: This flag will allow us to work with interactive processes by allocating a tty for the container.
+
+- `--privileged`: This flag gives the container access to the hosts' devices. This allows us to take advantage of certain system calls used by Wallaroo, specifically `mbind` and `set_mempolicy`. This flag is optional, but by excluding it there will be a performance degradation in Wallaroo's processing capabilities.
```
"this flag gives" -> "Give the ..."
"This allows us..." -> "This allows certain system calls to be used by ..."
"This flag is optional" -> "This setting is optional, but omitting it will result in a performance degradation."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `--privileged`: This flag gives the container access to the hosts' devices. This allows us to take advantage of certain system calls used by Wallaroo, specifically `mbind` and `set_mempolicy`. This flag is optional, but by excluding it there will be a performance degradation in Wallaroo's processing capabilities.
+
+- `-p 4000:4000`: This flag will map the default port for HTTP requests for the Metrics UI from the container to the host. This will make it possible to call up the Metrics UI from a browser on the host.
```
omit "This flag will"
"This will make it possible" -> "This makes it possible"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `-p 4000:4000`: This flag will map the default port for HTTP requests for the Metrics UI from the container to the host. This will make it possible to call up the Metrics UI from a browser on the host.
+
+- `-v /tmp/wallaroo-docker/wallaroo-src:/src/wallaroo`: This flag will mount a host directory as a data volume within the container. The first time you run this, you must use an empty directory in order for the Docker image to copy the Wallaroo source code to your host. This will allow you to open and modify the Wallaroo source code with the editor of your choice on your host. The Wallaroo source ...
```
Omit "this flag will"

nisanharamati replied:
"but without it you'd need" -> "but without it you will need"  (don't use a contraction here)

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `-v /tmp/wallaroo-docker/wallaroo-src:/src/wallaroo`: This flag will mount a host directory as a data volume within the container. The first time you run this, you must use an empty directory in order for the Docker image to copy the Wallaroo source code to your host. This will allow you to open and modify the Wallaroo source code with the editor of your choice on your host. The Wallaroo source ...
+
+- `-v /tmp/wallaroo-docker/python-virtualenv:/src/python-virtualenv`: This flag will mount a host directory as a data volume within the container. The first time this is run for the provided directory, this command will setup a persistent Python Virtual Environment for the container on your host. Thus, if you need to install any python modules using pip or easy_install they will persist after the ...
```
omit "this flag will"

nisanharamati replied:
if it's specifically using "virtualenv", then it is called "virtualenv" and not "Virtual Environment", and perhaps you should make the "virtualenv" name an href to https://virtualenv.pypa.io/en/stable/ (in the first occurence in this line)

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `-v /tmp/wallaroo-docker/python-virtualenv:/src/python-virtualenv`: This flag will mount a host directory as a data volume within the container. The first time this is run for the provided directory, this command will setup a persistent Python Virtual Environment for the container on your host. Thus, if you need to install any python modules using pip or easy_install they will persist after the ...
+
+- `--name wally`: The name for the container. This flag is optional but makes it easier to reference the container when we open other Bash sessions in the container.
```
"...easier to reference the container in later commands."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+Let's get started!
+
+## Terminal 1, Start the Wallaroo Docker container
```
"Start" -> "Starting"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- `--name wally`: The name for the container. This flag is optional but makes it easier to reference the container when we open other Bash sessions in the container.
+
+## Terminal 2, Start the Metrics UI
```
"Starting"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+docker exec -it wally environment-setup.sh
+```
+
+This command will start a new Bash shell within the container, we enter via the `environment-setup.sh` script to ensure we're in our persistent Python Virtual Environment.
````
"... within the container, which will run the `environment-setup.sh` script to ensure our persistent Python virtualenv is setup."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+metrics_reporter_ui stop
+```
+
+If you need to start the UI after stopping it, run:
````
not necessary, can be omitted

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+metrics_reporter_ui start
+```
+
+## Terminal 3, Run Giles Receiver
````
"Running"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+docker exec -it wally environment-setup.sh
+```
+
+We'll use Giles Receiver to listen for data from our Wallaroo application.
````
"We will"

nisanharamati replied:
Or better yet, "Giles Receiver will listen to..."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+receiver --listen 127.0.0.1:5555 --no-write --ponythreads=1 --ponynoblock
+```
+
+You should see the `Listening for data` that indicates that Giles receiver is running.
````
"You should see the line `Listening for data` ..."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+You should see the `Listening for data` that indicates that Giles receiver is running.
+
+## Terminal 4, Run the "Celsius to Fahrenheit" Application
```
"Running"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+docker exec -it wally environment-setup.sh
+```
+
+First, we'll need to get to the python Celsius example directory with the following command:
````
"we need"

nisanharamati replied:
or omit that, and just go with "First, get to the python..."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+cd /src/wallaroo/examples/python/celsius
+```
+
+Now that we are in the proper directory, and the metrics UI and a data receiver are up and running, we can run the application itself by executing the following command:
````
maybe just 
"Start the Celsius application by..."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+  --cluster-initializer --ponythreads=1 --ponynoblock
+```
+
+This tells the "Celsius to Fahrenheit" application that it should listen on port `7000` for incoming data, write outgoing data to port `5555`, and send metrics data to port `5001`.
````
"This tells" -> 

This command tells machida to:
* run the Celsius to Fahrenheit application
* listen on port 7000 for incoming data
* send output to port 5555
* send metrics data to port 5001

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+This tells the "Celsius to Fahrenheit" application that it should listen on port `7000` for incoming data, write outgoing data to port `5555`, and send metrics data to port `5001`.
+
+## Terminal 5
```
What are we going to do in this terminal? 
Should this be "Terminal 5, Sending data" ?

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+docker exec -it wally environment-setup.sh
+```
+
+### Sending Data with Giles Sender
````
If you change the "Terminal 5" heading above, you can probably omit this line

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+### Sending Data with Giles Sender
+
+We will be sending in 25,000,000 messages using a pre-generated data file. The data file will be repeatedly sent via Giles Sender until we reach 25,000,000 messages.
```
"We will send 25,000,000 messages using an included data file, which will be repeated by Giles Sender in order to reach 25,000,000 messages."

aturley replied:
I don't think this is much of an improvement on the original. In the original the word "repeatedly" is used, which indicates that something is done over and over again; the new one says "which will be repeated", which indicates that an action is done again, not done again and again. The new one also introduces some confusion about what is being "repeated".

You could try something like this:

> We will use Giles Sender to send 25,000,000 messages to the application. The messages come from an included data file, where each line in the file is used as a message. The sender loops over the file until it has sent 25,000,000 messages.

Then again, maybe it is sufficient to say:

> We will use Giles Sender to send 25,000,000 messages using an included data file.

And maybe provide a link to the Giles Sender documentation that the user can follow if they'd like to learn more about the specifics of how Giles Sender behaves.

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+We will be sending in 25,000,000 messages using a pre-generated data file. The data file will be repeatedly sent via Giles Sender until we reach 25,000,000 messages.
+
+You will now be able to start the `sender` with the following command:
```
"Start the `sender` with the following command:"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+If the sender is working correctly, you should see `Connected` printed to the screen. If you see that, you can be assured that we are now sending data into our example application.
+
+## Check Out Some Metrics
```
"Check the Metrics"

aturley replied:
"Check Out Some Metrics" is consistent with the section title in "Run a Wallaroo Application".

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+If your landing page resembles the one above, the "Celsius to Fahrenheit" application has successfully connected to the Metrics UI.
+
+Now, let's have a look at some metrics. By clicking on the "Celsius to Fahrenheit" link, you'll be taken to the "Application Dashboard" page. On this page you should see metric stats for the following:
```
"Now, lets..." -> "To look at the metrics, click on ..., which will take you to..."

nisanharamati replied:
"On this page you should see" -> "On this page, you will see"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+- a single pipeline: `Celsius Conversion`
+- a single worker: `Initializer`
+- three computations: `Add32`, `Decode Time in TCP Source`, `Multiply by 1.8`
```
Break into an intended bullet list?

e.g.
- three computations
    - `Add32`
    - `Decode...`
    ...

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+![Application Dashboard Page](/book/metrics/images/application-dashboard-page.png)
+
+You'll see the metric stats update as data continues to be processed in our application.
```
"The performance metrics update as new data is processed by the application."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+You'll see the metric stats update as data continues to be processed in our application.
+
+You can then click into one of the elements within a category, to get to a detailed metrics page for that element. If we were to click into the `Add32` computation, we'll be taken to this page:
```
"Click on an element within a category to get a..."
"If you click on the `Add32`..., you will be..."

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+![Computation Detailed Metrics page](/book/metrics/images/computation-detailed-metrics-page.png)
+
+Feel free to click around and get a feel for how the Metrics UI is setup and how it is used to monitor a running Wallaroo application. If you'd like a deeper dive into the Metrics UI, have a look at our [Monitoring Metrics with the Monitoring Hub](/book/metrics/metrics-ui.md) section.
```
omit "Feel free to"

nisanharamati replied:
"If you would like" (no contraction)

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+## Shutdown
+
+### Terminal 6, Cluster Shutdown
```
"Terminal 6, Shutting Down the Cluster"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+docker exec -it wally environment-setup.sh
+```
+
+You can shut down the cluster with this command once processing has finished:
````
"Shut the cluster down with this..."

is the "once processing has finished" part really necessary?

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+cluster_shutdown 127.0.0.1:5050
+```
+
+You can shut down Giles Sender and Giles Receiver by pressing Ctrl-c from their respective shells.
````
"in their respective" (not "from")

nisanharamati replied:
omit "You can"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,201 @@
+
+You can shut down Giles Sender and Giles Receiver by pressing Ctrl-c from their respective shells.
+
+You can shut down the Metrics UI with the following command:
```
Omit "You can"

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+docker stop wally
+```
+
+If you hadn't already disconnected from the other shells, this command will close all those sessions as well.
````
"This command will also terminate any active sessions you may have left open to the docker container."

nisanharamati commented:
@JONBRWN I went over the two new book sections, and most of my comments are about keeping the voice and tone of the text consistent with the rest of the docs. The suggestions are mainly aimed at changing the tone from "blog post" to "technical documentation".

The one other issue is the link in `SUMMARY.md` which has a typo in its destination address.

aturley commented on SUMMARY.md:
```diff
@@ -13,11 +13,14 @@
-  * [MacOS Installation](book/getting-started/macos-setup.md)
-  * [Ubuntu Installation](book/getting-started/linux-setup.md)
-* [Run a Wallaroo Application](book/getting-started/run-a-wallaroo-application.md)
+* First Install
```
I think that "First Install" might be a confusing term for this. If the user doesn't have the background for what's happening here they may not understand why there is a "first" install, if there will be other installs (second, third, etc), what this install involves, or why it is a step at all.

I'm not sure how to best address this. Part of me thinks that maybe this section should be called something like "Using the Docker Quickstart Environment" or something like that. I don't really feel great about that title, but I think it does a better job of expressing the intention here.

SeanTAllen replied:
Installation or Getting Started is much better than "First Install"

SeanTAllen commented:
Why would I use Docker? Why from source? I can't find any explanation of which to use. That's going to be confusing.

Sierra and High Sierra OSX folks should use Docker due to possible kernel panics. But that isn't mentioned.

SeanTAllen commented:
"Run A Application in Docker" <-- `an application` not `a application`

aturley commented on book/getting-started/run-a-wallaroo-application-docker.md:
````diff
@@ -0,0 +1,201 @@
+To shut down the Wallaroo container, use the `docker stop` command:
+
+```bash
+docker stop wally
````
It might be worth pointing out that this command should be run from your computer and not from within the Docker container. This feels like something that could be confusing to folks who are not used to working with Docker.

JONBRWN commented:
@SeanTAllen I've addressed the "Why" for using Docker and added a note in the Mac OS instructions as well. If I need additional information for why someone should install from source let me know

nisanharamati commented:
@JONBRWN you should add a link to `Tips for using docker...` in the `SUMMARY.md` file

nisanharamati commented on book/getting-started/run-a-wallaroo-application-docker.md:
```diff
@@ -0,0 +1,205 @@
+# Run an Wallaroo Application in Docker
```
'an' -> 'a'

JONBRWN replied:
@nisanharamati that's what I had before, then @SeanTAllen suggested:

"Run A Application in Docker" <-- an application not a application

One of you care to make the decision?

nisanharamati replied:
There are two different cases:
Run **a W**allaroo Application...
Run **An A**pplication...

SeanTAllen replied:
what nisan said. that's the correct grammar.

nisanharamati commented on book/getting-started/docker-setup.md:
```diff
@@ -0,0 +1,55 @@
+
+## What's Included in the Wallaroo Docker image
+
+### Machida
```
This might just be my personal aesthetic, but I think the H3 three headings look too large for this:
![image](https://user-images.githubusercontent.com/3074526/33733357-65d1ae26-db3e-11e7-97f0-4bedc8c54aeb.png)

Maybe it should be H4 or H5, or possibly just bold text, or a bullet list?

JONBRWN replied:
I'll go with bullet list, with item in bold

JONBRWN commented:
@nisanharamati @SeanTAllen @aturley @dipinhora if two people can approve their reviews I'll consider this good, squash the commits and merge this

JONBRWN commented:
thanks @SeanTAllen & @nisanharamati, I will squash and merge
